Free temporal envelope for the GaussianPulse profile, replaces ExpRampWithPrepulse

pordyna opened the pull request:
* Adds a user free functor for defining GaussianPulse temporal envelope (the transversal envelope remains Gaussian). 
* Also, removes ExpRampWithPrepulse from PIConGPU that could be used as an envelope for the GaussianPulse already before. Such complicated profiles can be now defined with the free functor. 
* Adds GaussianPulse laser profile to FoilLCT example together with a free functor example (that just reproduces the standard GaussianPulse). 

I tested this feature already in production, it works as expected. I believe @paschk31 was also already using it. 
@steindev @PrometheusPi @psychocoderHPC This is ready for review. The draft state is just so that I can remove the last commit before merging (it is there just so that we can test the examples with CI without running all possible tests).

pordyna committed:
Add FreeFunctor for GaussiaPulse temporal envelope

ci: no-compile

add free functor for gaussian envelope

be more descriptive in free formula laser temporal envelope docstrings

Move free temporal envelope to FoilLCT:

* add GaussianPulse to FoilLTC
* move GaussianPulseWithFreeEnvelope to FoilLTCT from LWFA example
* remove ExpRampWithPrepulse from FoilLCT

remove exponential ramp laser from LWFA example

remove ExpRampWithPrepulse

temporarily test all cmakeFlags in FoilLCT

fix undefined variable, clarify where 15 is coming from in the example

fix inconsistent laser profile naming in an example case

Revert "temporarily test all cmakeFlags in FoilLCT"

This reverts commit a8eaf4299e177791680ed9f6927e97f443412388.

remove extra documentation from incidentField template,that was added again by accident

document temporal envelope in incidentField.param docstring

Renaming and typos for Gaussian Free Envelope

rephrase dogygen comment for GaussianPulseParam

PrometheusPi commented:
Something failed in the CI but there was no trace, thus I restarted the failed CI jobs.

chillenzer commented:
@pordyna What's the status of this? Any movement?

PrometheusPi commented:
@pordyna Is this reviewable and the draft mode a left-over, or should we wait for more changes?

pordyna commented:
@PrometheusPi you can have a look already, but I still need to move the extra test out of the CI (was just temporary). Also, I will provide (not in picongpu code but will link it here) an example on how to reproduce the `ExpRampWithPrepulse`, and test it against the old implementation.  Should be done with it today.

pordyna commented:
@finnolec  Here is a gist on how to create the old `ExpRampWithPrepulse` temporal envelope using the Free Functor and `sympy`. https://gist.github.com/pordyna/3425d2a053bb052833717a8719ac78f3

chillenzer commented:
I love how you immediately ripped that feature from deep inside the guts of PyPIConGPU out of context built something useful out of it!

pordyna commented:
Comparing Electric field values between the simulation, that uses a `sympy` generated input, and the `sympy` expression evaluated in python. Orange: `sympy` E-field envelope (corrected for propagation distance), Blue: $$\mathrm{abs}(E_x)$$ at the focus position in the middle of the simulation box. All in units of a0. The input profile is the same as in the gist mentioned in my last comment. Fits quite well, I would say.

<img width="1244" height="714" alt="compare_sympy" src="https://github.com/user-attachments/assets/7724cd47-8f2d-49ad-ad52-d489d8fdc29a" />

pordyna commented:
@chillenzer @psychocoderHPC ready for review

PrometheusPi reviewed (dismissed):
looks good to me - I would like to test the regular Gauss case before merging

steindev commented on include/picongpu/fields/incidentField/profiles/GaussianPulse.def:
```diff
@@ -119,6 +119,27 @@ namespace picongpu::fields::incidentField::profiles
 
+    /** User temporal envelope Functor
+     *
+     * This functor describes the E field evelope (sqrt(I))
```
typo 'envelope'.
Please also add 'temporal' again, just to make this clear.

steindev replied:
Maybe like this: 'This functor describes the longitudinal E-field envelope (sqrt(I))'

pordyna replied:
lol I'm so sure I commented on this and changed it? This is some git magic haha. I will fix it again.

steindev commented on include/picongpu/fields/incidentField/profiles/GaussianPulse.hpp:
```diff
@@ -387,6 +387,22 @@ namespace picongpu::fields::incidentField
+        template<typename T_Func>
+        struct FreeEnvelope
+        {
+            static constexpr float_X TIME_SHIFT = 0.0_X;
```
What is this doing here? And is it shadowing the `float_64 TIME_DELAY` from the `BaseParam`?

pordyna replied:
Oh, this is part of the interface for the temporal envelopes. I think, it was there because `ExpRampWithPrepulse` defined one. I guess now we could get rid of it. But this is also not exposed to user. So, this is a coordinate transform between pic and some internal profile time.

For now, it is required here:
https://github.com/ComputationalRadiationPhysics/picongpu/blob/0d5821fe732c12942cc249476955a4f9425821f1/include/picongpu/fields/incidentField/profiles/GaussianPulse.hpp#L226

steindev replied:
In the case of not exposed and not used (since always zero)we should get rid of it.

pordyna replied:
Offline discussion with @steindev: It is used here https://github.com/ComputationalRadiationPhysics/picongpu/blob/0d5821fe732c12942cc249476955a4f9425821f1/include/picongpu/fields/incidentField/profiles/GaussianPulse.hpp#L375 and it affects not only the envelope but also the phase. We decided that we should keep it.

steindev commented on share/picongpu/examples/FoilLCT/include/picongpu/param/incidentField.param:
```diff
@@ -214,90 +213,69 @@ namespace picongpu
 
-            //! Laser Profile is taking plane wave param configuration and is overwriting only a few settings
-            struct ExpRampWithPrepulseParam : public PlaneWaveParams
+            //! Extends plane wave setup with a gassian pulse paramters
```
Typo 'Gaussian' & 'parameters'.
Why is the word 'parameter' written here?

pordyna replied:
>  Why is the word 'parameter' written here?

Since it is just the param class. What would be your suggestion?

Fixed the typos

steindev replied:
Maybe: Extend PlaneWave setup to Gaussian pulse

steindev commented on share/picongpu/examples/FoilLCT/include/picongpu/param/incidentField.param:
```diff
@@ -214,90 +213,69 @@ namespace picongpu
-            //! Laser Profile is taking plane wave param configuration and is overwriting only a few settings
-            struct ExpRampWithPrepulseParam : public PlaneWaveParams
+            //! Extends plane wave setup with a gassian pulse paramters
+            struct GaussianPulseParam : public PlaneWaveParams
```
Why does this Gaussian param struct inherit from a plane wave param and does not use a normal Gaussian param struct?

pordyna replied:
`PlaneWaveParam` is just `BaseParam` in that context. Just reusing the parameters.

steindev commented on share/picongpu/examples/FoilLCT/include/picongpu/param/incidentField.param:
```diff
@@ -214,90 +213,69 @@ namespace picongpu
+             * 1.0 corresponds to the a_0 set in the gaussian param struct
+             * This functor reproduces the standard Gaussian envelope
+             */
+            struct FreeEnvelopeExample
```
The typical convention for the name would be `FreeEnvelopeFunctor`, wouldn't it?

pordyna replied:
corrected

steindev reviewed (commented):
So far so good. Only minor comments. but it is a pity, that the Exponential Ramp Profile or `GaussianBeamWithContrastCurve` is gone as an example from the source code.

Would you add your gist, which showcases the construction of a complex profile, to readthedocs? I think this would aid the loss :-)

pordyna commented:
@steindev I addressed your comments.

steindev commented:
@pordyna I responded to your two comments. Also, it seems you forgot to address the first of the initial comments.

Please squash and rebase before the next push. Then we can directly merge. Thanks!

pordyna commented:
Squashed commits and rebased on dev. Also addressed remaining @steindev comments (only affects docstrings). Ready to merge from my side. @PrometheusPi do you still want to run your test?

PrometheusPi commented:
I submitted the test today to hemera. Will check them as soon as they finished.

PrometheusPi commented:
There is no deviation in field energy components between the current dev and this PR. 
Can be merged from my side.
